port/osx/omrintrospect.c

```diff
@@ -353,16 +433,29 @@ omrintrospect_threads_startDo_with_signal(struct OMRPortLibrary *portLibrary, J9
 	data->state = state;
 	data->filterThread = mach_thread_self();
 
+	/* Create a pipe to allow the resumed thread to report it has completed
+	 * sending its callstack info.
+	 */
+	result = pipe(pipeFileDescriptor);
+	if (0 != result) {
+		RECORD_ERROR(state, INITIALIZATION_ERROR, result);
+		goto cleanup;
+	}
+	/* Initialize pipe to be non-blocking as poll() ensures read() does not block */
+	flag = fcntl(pipeFileDescriptor[0], F_GETFL);
+	fcntl(pipeFileDescriptor[0], F_SETFL, flag | O_NONBLOCK);
+
 	/* Suspend all threads bar this one. */
-	if (0 != suspendAllPreemptive(state->platform_data)) {
-		RECORD_ERROR(state, SUSPEND_FAILURE, suspend_result);
+	result = suspendAllPreemptive(state->platform_data);
+	if (0 != result) {
+		RECORD_ERROR(state, SUSPEND_FAILURE, result);
 		goto cleanup;
 	}
 
 	result = setupNativeThread(state, signal_info);
 	if (0 != result) {
 		RECORD_ERROR(state, COLLECTION_FAILURE, result);
-		goto cleanup;
+		return NULL;
```

Comment:
I believe I made this change because `setupNativeThread` only works on threads individually so if there is a problem with this first (current) thread, `nextDo` can continue iterating the rest of the threads. In `javadump.cpp` `startDo` returning NULL does not terminate the thread collection, it tries again (`omrthread_get_osId()` and a separate call to `j9introspect_backtrace_thread()`) then continues with `protectedNextDo`. Only `nextDo` returning NULL terminates thread collection.

Reply:
The problem is that the contract for this function is that 'nextDo' should only be called after this returns a non-NULL value.

Reply:
Ok, makes sense rereading the function description, will change. `nextDo` will return upon `(NULL == data)`